### PR TITLE
Fix typos, grammar, spelling and factual mistakes in the documentation

### PR DESCRIPTION
**Pull request by IAmMarcelJung**
After stumbling over another small mistake in the docmentation, I decided to check the whole documentation using [ltex](https://valentjn.github.io/ltex/). As the name implies, it is normally used for L<sup>A</sup>T<sub>E</sub>X documents, but it can also be used for  reStructuredText  (and others).  I did of course check the suggestions and only applied the ones that made sense to me. There is of course also no guarantee that every error was catched, but I am quite happy with what it found.

In `simulation.rst` there also was a factual error, stating that `0xFAB0` causes sampling the data and creating a one-cycle strobe, although actually `0xFAB1` causes it.

As always, any comments or suggestions are appreciated!

### CODE REVIEW

**Comment by IAmMarcelJung** [comment]
I found even more mistakes and will include them in another commit.

**Comment by IAmMarcelJung** [comment]
Although the generation runs without errors, the generated files seem to be broken. I will investigate it and remove the draft if it is solved.

**Comment by IAmMarcelJung** [comment]
I messed up my local sphinx install. After repairing it, I could generate valid HTML files again and view the results. The changes can now be reviewed.

**Comment by KelvinChung2000** [comment]
I will try to review them tomorrow. 

Since you are working on the doc, it would be nice if you reviewed #197, which greatly enhances its appearance. 

Probably not within this pull request, but it would also be nice if you could copy some of the yosys and nextpnr doc to FABulous doc to make FABulous more approachable for people not familiar with the Linux like #205

File: docs/source/FPGA-to-bitstream/VPR compilation.rst
Comment (KelvinChung2000) [inline docs/source/FPGA-to-bitstream/VPR compilation.rst]: Is this supposed to use the arrow symbol?

File: docs/source/simulation/simulation.rst
Comment (KelvinChung2000) [inline docs/source/simulation/simulation.rst:1]: I will assume you have confirmed this to be true.

**Comment by IAmMarcelJung** [comment]
Thanks! I am not actively working on the documentation, this was mainly done for procastination...

However, I can still try improve the documentation and make it more beginner friendly. To my shame, I think I also did not document  #138, so this also has to be done.

When I worked with the documentation, I noticed that the files are not formatted at all and some files had exceptionally long line, which made it a bit tedious to review them. Since we now use black as a formatter for python code, I looked for a restucturedText formatter. I only found [rstfmt](https://github.com/dzhu/rstfmt) and [docstrfmt](https://github.com/LilSpazJoekp/docstrfmt). They both seem to be more or less active and the latter is influenced by the former. Should I open a discussion about if and which formatter to use?

**Comment by KelvinChung2000** [comment]
I also should be working on something other than FABulous at the moment...  procrastination...

More formatting would be nice. Since you are on the boat now, you can make the decision. Better doc support is a continuous effort, so do them when you don't want to work 😉. To be honest, once we moved to pip setup, things should become simpler. If the setup tool can also install yosys and nextpnr, then we don't even need to make doc for them.

**Comment by IAmMarcelJung** [comment]
Okay then I don't think it is necessary to document the installation of yosys and nextpnr any further. 

I forgot to mention that I tried out docstrfmt, but it broke my HTML documents, although they were generated without any error. To get the generation working, I also had to change the sphinx version, so maybe it will also be fixed once #197 is done. I will definietly wait with the formatting until #197 and this PR are merged, since it might get confusing otherwise.

Good to know I'm not the only one who is procratsitnating 😄

File: docs/source/FPGA-to-bitstream/Nextpnr compilation.rst
Comment (TaoBi22) [inline docs/source/FPGA-to-bitstream/Nextpnr compilation.rst]: I'm not sure these lines are super clear anyway. Maybe something more descriptive like

```suggestion
Nextpnr can compile a JSON description of a circuit to FASM (the FPGA Assembly format, describing a concrete list of features on a specific FPGA fabric to be enabled) using the architectural description in bels.txt and pips.txt
```
would be nice. Also don't think it would be entirely unreasonable to drop these lines given it's all explained below - I'd guess they were originally added as placeholders by Jing a few years ago while he fleshed out the docs (and git blame seems to agree)

File: docs/source/FPGA-to-bitstream/VPR compilation.rst
Comment (TaoBi22) [inline docs/source/FPGA-to-bitstream/VPR compilation.rst]: As above, maybe worth making this clearer - something like

```suggestion
VPR can compile a circuit in BLIF format to FASM (the FPGA Assembly format, describing a concrete list of features on a specific FPGA fabric to be enabled) using the architectural description in architecture.xml and routing_resources.xml
```

File: docs/source/fabric_definition.rst
Comment (TaoBi22) [inline docs/source/fabric_definition.rst]:
```suggestion
 The multiplexers in :ref:`primitives` can either be controlled by configuration bits (e.g., to select if a LUT output is to be routed to a primitive output pin or through a flop) or by the user logic (e.g., to cascade adjacent LUTs for implementing larger LUTs, like the F7MUX and F8MUX multiplexers in Xilinx FPGAs with LUT6s).
```

File: docs/source/fabric_definition.rst
Comment (TaoBi22) [inline docs/source/fabric_definition.rst]: This is a weirdly inlined TODO anyway

```suggestion
  * ``SHARED_PORT``: this directive can only be used together optionally with ``EXTERNAL``.
  
TODO - shared  ports flagged with this directive are not connected to the switch matrix but are exported through the tile entity to the top-level fabric wrapper.
```

**Review by TaoBi22 — APPROVED** [review]
My obligatory drive-by of anything to do with docs :)

Thanks for looking at this!

File: docs/source/FPGA-to-bitstream/VPR compilation.rst
Comment (IAmMarcelJung) [reply docs/source/FPGA-to-bitstream/VPR compilation.rst]: Yeah, this should have been it, but when I built it, it didn't work, so I replaced it in a more recent commit. At least, that's what I thought. I'm currently not working, but I will check it once I get back to it. Thanks for pointing it out!

File: docs/source/FPGA-to-bitstream/Nextpnr compilation.rst
Comment (IAmMarcelJung) [reply docs/source/FPGA-to-bitstream/Nextpnr compilation.rst]: So what do you think, should we still include it or remove it completely? If so, we should either link to FASM or create a footnote since the text in the bracket breaks the reading flow a bit IMHO.

File: docs/source/simulation/simulation.rst
Comment (IAmMarcelJung) [reply docs/source/simulation/simulation.rst:1]: No this was just a formulation fix, I assumed myself that it is true. Should we remove this? I don't know about the state of the emulation setup.

File: docs/source/FPGA-to-bitstream/VPR compilation.rst
Comment (IAmMarcelJung) [reply docs/source/FPGA-to-bitstream/VPR compilation.rst]: Same as above, do we keep it with the new formulation or remove it? I think it's a good summary of what VPR does from a user perspective.

Comment (IAmMarcelJung) [reply docs/source/FPGA-to-bitstream/VPR compilation.rst]: Fixed, thanks again for spotting it!

File: docs/source/fabric_definition.rst
Comment (IAmMarcelJung) [reply docs/source/fabric_definition.rst]: Thanks for this! I will add it to a batch commit once the other discussions are resolved.

File: docs/source/fabric_definition.rst
Comment (IAmMarcelJung) [reply docs/source/fabric_definition.rst]: I was also thinking about using the [Sphinx builtin TODO](https://www.sphinx-doc.org/en/master/usage/extensions/todo.html#directive-todolist). Maybe this would be a better option?

Comment (TaoBi22) [reply docs/source/fabric_definition.rst]: I would say so!

File: docs/source/FPGA-to-bitstream/Nextpnr compilation.rst
Comment (TaoBi22) [reply docs/source/FPGA-to-bitstream/Nextpnr compilation.rst]: There's not really a proper breakdown on the page, so I think it would be reasonable to keep it in this case (I'll comment separately on the other case) - a footnote sounds good!

File: docs/source/FPGA-to-bitstream/VPR compilation.rst
Comment (TaoBi22) [reply docs/source/FPGA-to-bitstream/VPR compilation.rst]: I would say the paragraph directly below it is already a decent summary of the page, so the two paragraphs are just repeating each other at the moment. Could drop one or merge them, up to you!

Comment (IAmMarcelJung) [reply docs/source/FPGA-to-bitstream/VPR compilation.rst]: Yeah you're right, only the `BLIF` is missing. I will add it to the exisiting paragraphs.

File: docs/source/fabric_definition.rst
Comment (IAmMarcelJung) [reply docs/source/fabric_definition.rst]: I will use it then!

File: docs/source/simulation/simulation.rst
Comment (KelvinChung2000) [reply docs/source/simulation/simulation.rst:1]: I am also not sure about the emulation setup. I know the generated Verilog has something related to emulation, but I don't know how to use them.

File: docs/source/fabric_definition.rst
Comment (IAmMarcelJung) [reply docs/source/fabric_definition.rst]: After using the todo extenstion, it looks like this:

image

Should the to-do be included in the text at all? Or should we just use the todo extension to have the todos in the source files but not in the actual documentation? If we want to have it in the actual documentation, we should then also create a separate page for the to-dos. I personally would just keep it inside the source and hide it in the docs since it looks to much like work in progress if you see a to-do somewhere. But this is of course debatable :)

File: docs/source/simulation/simulation.rst
Comment (IAmMarcelJung) [reply docs/source/simulation/simulation.rst:1]: Then it might be reasonable to leave some kind of hint that it it is not implemented completly/100% tested/100 % stable/etc. I will ask Dirk what he knows about the emulation setup.

File: docs/source/fabric_definition.rst
Comment (TaoBi22) [reply docs/source/fabric_definition.rst]: Hmm yeah, this is a little in your face. Maybe the best option is actually just to comment it out (especially since the VPR flow might end up deprecated anyway)?

Comment (TaoBi22) [reply docs/source/fabric_definition.rst]: Oh, maybe I misunderstood - do you mean Sphinx has an option to hide the TODOs in the actual builds? That also works

Comment (IAmMarcelJung) [reply docs/source/fabric_definition.rst]: > Hmm yeah, this is a little in your face. Maybe the best option is actually just to comment it out (especially since the VPR flow might end up deprecated anyway)

We have three other TODOs (all in `fabric_definition.rst`). I would apply the same method to all of them to keep it consitent.

> Oh, maybe I misunderstood - do you mean Sphinx has an option to hide the TODOs in the actual builds? That also works

The TODOs are actually hidden by default (see also this [SO discussuion](https://stackoverflow.com/questions/50314856/enabling-the-todo-feature-in-sphinx-docs)).  There is also the `todo_emit_warnings` option which should show warnings when building, but it is not yet working for me. This would IMHO be the best approach since TODOs can easily be forgotten over time ;)

Comment (TaoBi22) [reply docs/source/fabric_definition.rst]: Ah, ok, great! Agreed then, if the TODO blocks aren't shown by default then wrapping things up in them sounds good! 

I would hold off on `todo_emit_warnings` for now as the ReadTheDocs build is configured to fail on warnings, so this will cause all our docs to stop building. Could always be something to change later and modify the RTD config though.

Comment (IAmMarcelJung) [reply docs/source/fabric_definition.rst]: I did manage to make the warnings appear. I did not change the source file after adding `todo_emit_warnings` and therefore the file was not rebuilt..
`make clean` to the rescue :) 

> [...] the ReadTheDocs build is configured to fail on warnings, so this will cause all our docs to stop building [...]

I assumed that something like this would be the case. Then I will leave it for now of course.

**Comment by IAmMarcelJung** [comment]
The last commit (d7cf183) implements using the todo extension. Now no TODO is in the generated docs anymore.

Also two TODOs are finished:
1. Cite the "Shrink my FPGAs" paper where it was appropriate (I asked Dirk for this)
2. Add a description for `SHARED_PORT`. An additional example of a `SHARED_PORT` was  added in both the Verilog and the VHDL examples (Dirk told me to do so and I also think it is useful).

Two "bug fixes: were made:
1. Changed Out_Pad to eFPGA for at the end of the eFPGA entity.
2. Add `end entity LUT4` in the example in the Bitstream remapping chapter.

I also asked Dirk for the other TODOs, but I'm still waiting for the answer.

If anyone else knows what has to be written for the TODOs you can of course also let me know or directly change it (I think only short additions are needed. Thanks in advance!

**Comment by IAmMarcelJung** [comment]
In the latest commits, another TODO could be fixed (thanks to @EverythingElseWasAlreadyTaken!). Also a small formulation and punctuation fix was done.

File: docs/source/simulation/simulation.rst
Comment (IAmMarcelJung) [reply docs/source/simulation/simulation.rst:1]: I recently asked Dirk and he told me what to right. I will add it in a commit and mark this PR as ready for review.

**Comment by IAmMarcelJung** [comment]
In the last commit (0b7f590) just the hint on the emulation setup was updated and turned into a note, so it does not look anymore as if it was just some internal comment.

**Review by KelvinChung2000 — APPROVED** [review]
Should I merge this first or #210 ? Since the `conf.py` might lead to merge conflict.

**Comment by IAmMarcelJung** [comment]
Will it also lead to conflicts if I revert the changes made by black? Other than that I  just added the todo extension. I could just revert the formatting if that is enough to avoid conflicts.

**Comment by KelvinChung2000** [comment]
I thought this is going to development, never mind.

**Pull request merged**
